### PR TITLE
Define encoding/decoding for module imports

### PR DESCRIPTION
**Pull request by akrmn**
This defines the encoding/decoding of imports (specified by packageId + module name) as the `$$imports` value in DAML-LF

First step towards closing #10773

changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion/MetadataEncoding.hs [hunk 1]
```diff
@@ -153,6 +158,53 @@ decodeOverlapMode = \case
...
+decodeModuleImports = fmap S.fromList . decodeTypeList decodeModuleImport
+
+decodeModuleImport :: LF.Type -> Maybe (LF.Qualified ())
+decodeModuleImport x = decodeTypeList Just x  >>= \case
+    [packageRef, moduleName] ->
+        LF.Qualified
+            <$> decodePackageRef packageRef
+            <*> decodeModuleName moduleName
+            <*> pure ()
+    _ -> Nothing
```
Comment (sofiafaro-da) [inline compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion/MetadataEncoding.hs:196]: Just a suggestion, but I think this benefits from the `MonadFail` instance for `Maybe`:

```suggestion
decodeModuleImport x = do
    [p, m] <- decodeTypeList Just x
    packageRef <- decodePackageRef p
    moduleName <- decodeModuleName m
    pure (LF.Qualified packageRef moduleName ())
```

File: compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion/MetadataEncoding.hs [hunk 2]
```diff
@@ -153,6 +158,53 @@ decodeOverlapMode = \case
...
+decodePackageRef :: LF.Type -> Maybe LF.PackageRef
+decodePackageRef = \case
+    LF.TUnit -> pure LF.PRSelf
+    TEncodedStr packageId -> pure (LF.PRImport (LF.PackageId packageId))
+    _ -> Nothing
+
+decodeModuleName :: LF.Type -> Maybe LF.ModuleName
+decodeModuleName = fmap LF.ModuleName . decodeTypeList \case
+    TEncodedStr component -> pure component
+    _ -> Nothing
```
Comment (sofiafaro-da) [inline compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion/MetadataEncoding.hs:207]: The `decodeList \case ...` is a bit hard to parse ... I think it would be nice to factor the `\case ...` part out into a separate function, `decodeText :: LF.Type -> Maybe T.Text`

**Review by sofiafaro-da — APPROVED** [review]
Looks great, thank you!

**Review by cocreature — APPROVED** [review]
Nice work, thanks!

File: compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion/MetadataEncoding.hs (hunk 1 above)
Comment (akrmn) [reply compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion/MetadataEncoding.hs:196]: ah nice! thanks :)
